File: 0987-vertical-order-traversal-of-a-binary-tree/0987-vertical-order-traversal-of-a-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    class Element{
        TreeNode node;
        int col;
        int row;

        Element(TreeNode node,int col,int row){
            this.node = node;
            this.col = col;
            this.row = row;
        }
    }
    public List<List<Integer>> verticalTraversal(TreeNode root){
        List<List<Integer>> result = new ArrayList<>();

        if(root == null){
            return result;
        }
        Queue<Element> queue = new LinkedList<>();
        List<int[]> elements = new ArrayList<>();
        int minVal = 0;
        int maxVal = 0;
        queue.add(new Element(root,0,0));
        while(!queue.isEmpty()){
            Element ele = queue.remove();
            minVal = Math.min(minVal,ele.col);
            maxVal = Math.max(maxVal,ele.col);
            elements.add(new int[]{ele.col,ele.row,ele.node.val});
            if(ele.node.left != null) 
                queue.add(new Element(ele.node.left,ele.col-1,ele.row+1));
            if(ele.node.right != null)
                queue.add(new Element(ele.node.right,ele.col+1,ele.row+1));
            


        }
        Collections.sort(elements,(a,b)->{
            if(a[0] != b[0])
                return a[0] - b[0];
            if(a[1] != b[1])
                return a[1] - b[1];
            return a[2] - b[2];

        });
        int prevCol = Integer.MIN_VALUE;
        for(int[] temp : elements){
            int col = temp[0];
            int val = temp[2];
            if(col != prevCol){
                result.add(new ArrayList<>());
                prevCol = col;

            }
            result.get(result.size()-1).add(val);


        }
        return result;








      }
}

File: 0987-vertical-order-traversal-of-a-binary-tree/0987-vertical-order-traversal-of-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from collections import deque
class Solution:
    def verticalTraversal(self, root: Optional[TreeNode]) -> List[List[int]]:
        result = []
        temp = []
        minVal  = 0 
        if not root:
            return result
        queue = deque()
        queue.append([0,0,root])
        while queue:
            ele = queue.popleft()
            if ele[2].left:
                queue.append([ele[0] - 1, ele[1] + 1, ele[2].left])
            if ele[2].right:
                queue.append([ele[0] + 1 , ele[1] + 1 , ele[2].right])
            temp.append(ele)
            minVal = min(minVal,ele[0])
        
        def criteria(x):
            return (x[0],x[1],x[2].val)



        temp.sort(key = criteria)
        prev = float("-inf")

        
        cur = 0
        for i in temp:
            if i[0] != prev:
                result.append([])
                prev = i[0]
            result[-1].append(i[2].val)
        

        return result
        
            

            

            

            

        
        
        